src/client/dns.rs

```diff
@@ -6,27 +6,42 @@ use std::net::{
 };
 use std::vec;
 
-use ::futures::{Async, Future, Poll};
+use futures::{Async, Future, Poll};
+use futures::task;
+use futures::future::lazy;
+use futures::executor::Executor;
+use futures::channel::oneshot;
 
-pub struct Work {
-    host: String,
-    port: u16
+pub struct Resolving {
+    receiver: oneshot::Receiver<Result<IpAddrs, io::Error>>
 }
 
-impl Work {
-    pub fn new(host: String, port: u16) -> Work {
-        Work { host: host, port: port }
+impl Resolving {
+    pub fn spawn(host: String, port: u16, executor: &mut Executor) -> Resolving {
+        let (sender, receiver) = oneshot::channel();
+        executor.spawn(Box::new(lazy(move |_| {
+            debug!("resolving host={:?}, port={:?}", host, port);
+            let result = (host.as_ref(), port).to_socket_addrs()
+                .map(|i| IpAddrs { iter: i });
+            sender.send(result).ok();
+            Ok(())
+        }))).unwrap();
```

Comment:
It looks like the receiver will return an error if the sender was dropped, so we should be able to just ignore this spawn error, right?